benchmarks/halo_finder/halo_finder.cpp

```diff
@@ -0,0 +1,269 @@
+  return out;
+}
+
+template <typename View>
+struct Wrapped
+{
+  View _M_view;
+  double _r;
+};
+
+template <typename View>
+auto wrap(View v, double r)
+{
+  return Wrapped<View>{v, r};
+}
+
+namespace ArborX
+{
+namespace Traits
+{
+template <typename View>
+struct Access<Wrapped<View>, PredicatesTag>
+{
+  using memory_space = typename View::memory_space;
+  static size_t size(Wrapped<View> const &w) { return w._M_view.extent(0); }
+  static KOKKOS_FUNCTION auto get(Wrapped<View> const &w, size_t i)
+  {
+    return intersects(Sphere{w._M_view(i), w._r});
+  }
+};
+} // namespace Traits
+} // namespace ArborX
+
+template <typename DeviceType>
+bool verifyHalos(Kokkos::View<int *, DeviceType> offset,
+                 Kokkos::View<int *, DeviceType> indices,
+                 Kokkos::View<int *, DeviceType> halos)
+{
+  using ExecutionSpace = typename DeviceType::execution_space;
+  int num_nodes = halos.size();
+  ARBORX_ASSERT(offset.size() == num_nodes + 1);
+  ARBORX_ASSERT(ArborX::lastElement(offset) == indices.size());
+
+  // Check that there are no negative halo indices
+  int num_incorrect = 0;
+  Kokkos::parallel_reduce("verify_negative_indices",
+                          Kokkos::RangePolicy<ExecutionSpace>(0, num_nodes),
+                          KOKKOS_LAMBDA(int i, int &update) {
+                            if (halos(i) < 0)
+                            {
+                              printf("Negative halo index: %d\n", i);
+                              update++;
+                            }
+                          },
+                          num_incorrect);
+  if (num_incorrect)
+    return false;
+
+  // Check that connected vertices have the same halo index
+  num_incorrect = 0;
+  Kokkos::parallel_reduce(
+      "verify_connected_indices",
+      Kokkos::RangePolicy<ExecutionSpace>(0, num_nodes),
+      KOKKOS_LAMBDA(int i, int &update) {
+        for (int j = offset(i); j < offset(i + 1); ++j)
+        {
+          if (halos(i) != halos(indices(j)))
+          {
+            printf("Non-matching halo indices: %d [%d] -> %d [%d]\n", i,
+                   halos(i), indices(j), halos(indices(j)));
+            update++;
+          }
+        }
+      },
+      num_incorrect);
+  if (num_incorrect)
+    return false;
+
+  // Check that non-connected vertices have different halo indices
+  auto halos_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, halos);
+  auto offset_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, offset);
+  auto indices_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, indices);
+
+  std::set<int> unique_halo_indices;
+  for (int i = 0; i < num_nodes; i++)
+    unique_halo_indices.insert(halos_host(i));
+  auto num_unique_halo_indices = unique_halo_indices.size();
+
+  int num_halos = 0;
+  std::set<int> halo_sets;
+  for (int i = 0; i < num_nodes; i++)
+  {
+    if (halos_host(i) >= 0)
```

Comment:
You checked for negative indices in the `parallel_reduce` above.

Reply:
No, this is different. We modify the indices to become negative inside the loop to avoid checking the same index twice.